### ISSUE TITLE
In Shipping builds, Cesium for Unreal ends up using UE's SQLite, causing crashes

### ISSUE THREAD

**Issue by kring** [CesiumGS/cesium-unreal#245]
@AlbanBERGERET-Epic reported that running packaged/shipping game while the editor is running consistently results in a crash and corruption of the cache database:

![image](https://user-images.githubusercontent.com/924374/111706272-1bee3500-8896-11eb-98b2-3973a4662674.png)

I wasn't able to reproduce this exactly, possibly because Alban is using an older version of Cesium for Unreal, or possibly due to some other subtle difference between our systems. I think we're catching the exception above in the latest versions. However, I do see lots of errors in the console:

`LogCesium: Error: [2021-03-19 09:28:18.281] [error] [SqliteCache.cpp:449] database is locked`

I don't think we want a separate cache database per application, but maybe there's a better place to put the cache DB than `FPaths::EngineUserDir`. 

The SQLite docs (https://www.sqlite.org/faq.html#q5) say it should be ok to share a database between processes as long as its on a real file system, not an NFS or Windows share. However, the DB will return SQLITE_BUSY. We should probably handle that by retrying.

However, based on the error message in the logs it seems like we're getting SQL_LOCKED instead, which seems weird: https://www.sqlite.org/rescode.html#locked

Long story short, we should investigate whether we're doing anything dodgy here.

**Comment by AlbanBERGERET-Epic** [comment]
@kring I found out more details about this one. 

Even with using your latest release binaries (v1.1.1, added directly to the engine), I still have this issue, but only when packaging a SHIPPED build. 
All is fine when packaging is done in development, but that should not be the final target. 

I hope it will help you to track it down...

**Comment by kring** [comment]
Thanks @AlbanBERGERET-Epic. I suspect the problem is that exceptions are disabled in Shipping builds, so the "database is locked" exception is causing the process to die instead.

**Comment by kring** [comment]
CC CesiumGS/cesium-native#212

**Comment by javagl** [comment]
Some (unfortunately not so helpful) insights:

- The error message "database is locked" indicates a `SQLITE_BUSY`. For a `SQLITE_LOCKED`, it would be different:
 
      /* SQLITE_BUSY        */ "database is locked",
      /* SQLITE_LOCKED      */ "database table is locked",

- According to the FAQ https://www.sqlite.org/faq.html#q5 this is anticipated for our case: 
  > When SQLite tries to access a file that is locked by another process, the default behavior is to return SQLITE_BUSY. 
  
- The "simple" approach of setting a "busy timeout" ( https://www.sqlite.org/c3ref/busy_timeout.html ) did not solve the problem for me
- (The same for using a pragma instead ( https://www.sqlite.org/pragma.html#pragma_busy_timeout ) - yeah, I was desperate and gave it a try...)
- I even registered a dedicated busy handler, as in https://www.sqlite.org/c3ref/busy_handler.html . It was occasionally triggered, and printed a log message "Tried it 1 times" ... up to "Tried it 30 times", and then apparently managed to perform the respective operation, but it sometimes still printed the "database is locked" error.

The callback was implemented to *always returned a non-zero value*, and thus, should actually cause it to retry the operation indefinitely. The fact that it still sometimes failed might mean that there is another limiting mechanism behind that, or that the busy handler is just not triggered in some cases. 

I might investigate that further in the future, but don't have further ideas about how to tackle this right now.

**Comment by baothientran** [comment]
Steps to reproduce the crash:

- Choose shipping configuration before package the project in the editor: File-> Package Project-> Build Configuration -> Shipping
- Package the project: File-> Package Project -> Windows
- Close the editor
- Run the package as a standalone app: Go to the package directory specified in the previous step -> WindowsNoEditor -> CesiumForUnrealSample.exe
- App crash

**Comment by baothientran** [comment]
Another progress on this issue is that for some reasons, in only shipping package, Unreal decides to use its own Sqlite3 plugin (which was enabled by other third party plugins before) to link against our `Sqlite3Cache` class, instead of using our libsqlite3 that is installed in the `cesium-unreal/Source/ThirdParty`. It doesn't happen with other configurations. After I disable the Sqlite3 plugin, I can't see any crash happening.

**Comment by kring** [comment]
Oh wow that's really interesting. And I guess the UE version of SQLite is either old or not built with the compile-options we expect/need?

**Comment by baothientran** [comment]
`sqlite3PagerWalSupported()` around line 59043 in the file `sqlite3.c` return false only for packaged game, so I think that confirms why database file created by the editor will crash the packaged game if it tries to access it

**Comment by kring** [comment]
@baothientran I followed your procedure above exactly, but I didn't see a crash when I ran the Shipping build (it worked just fine). 

However, the SQLite and SQLite Support plugins are not enabled on my system. Once I explicitly enabled them and repackaged, the app crashes on startup.

Presumably you have some other plugin enabled that enabled the SQLite one on your system? In any case, that helps explain why complaints about this aren't extremely wide-spread.

**Comment by baothientran** [comment]
@kring yup it only happens with sqlite plugin enable

**Comment by kring** [comment]
Just to confirm the obvious... When doing a shipping build, Unreal creates a massive link command-line with about a thousand things to be linked together. One of them is `SQLiteEmbedded.c.obj` which contains UE's custom SQLite. Another, much later on the link line, is our sqlite3.lib (built as part of cesium-native). This is why there's no multiply-defined symbol error. The linker finds the necessary symbols in an OBJ on the link line, and never bothers to look in our .lib.

I can really only see two paths forward:
1. Make do with UE's SQLite, despite its limitations.
2. Rename all the symbols in ours somehow so that we can use our own custom one.

**Comment by baothientran** [comment]
I feel like only option 1 is only a viable path for this (or advise everyone to turn off sqlite3 plugin before packaging) as option 2 requires a maintenance of a custom sqlite3 source code, so it's not really practical.

**Comment by kring** [comment]
I did a quick search/replace of `sqlite3` with `cqlite3` in sqlite3.h, sqlite3.c, SqlCache.h, and SqlCache.cpp. Plus renamed sqlite3.h/.c to cqlite3.h/.c. And it worked just fine. Total hack, but an easy one. I'm trying to think of something better.

Option 1 might be a good option, too. I don't totally understand the implications of that yet.

**Comment by kring** [comment]
# Implications of using UE's SQLite

If we use UE's SQLite, we can't have multiple simultaneous connections to the database, whether from the same process (i.e. multiple threads) or from multiple processes. This is because simultaneous connections require SQLite VFS support that UE doesn't implement.

We're not currently using multiple connections. Instead, our readers and writers - which operate from multiple threads simultaneously - all share a single database connection. This scenario will still work in UE's SQLite, too. SQLite effectively uses a mutex to serialize all DB access in this scenario, whether for read or write. We also have our own (partially redundant) mutex that does the same. However, we're probably leaving quite a lot of performance on the table by our choice not to use multiple connections. And by using UE's SQLite, we'd be unable to ever change this.

With UE's SQLite, we also can't use Write-ahead logging (WAL) unless we _also_ use `EXCLUSIVE` locking (https://sqlite.org/wal.html#use_of_wal_without_shared_memory). WAL is potentially faster, but the most important advantage of WAL is that when you have separate reader and writer connections, the readers and writers don't block each other. But since we're not using multiple connections anyway (see above), and `EXCLUSIVE` mode _prevents_ the use of multiple connections, the point is a bit moot.

For backward compatibility: to open an existing database set to use WAL mode, we'll have to set EXCLUSIVE locking. Or delete the DB and start over.

# Shared access between processes

The original problem described in this issue had to do with running the Editor and a Shipping game simultaneously, with both accessing the database. In theory, with a fully-functional SQLite (not UE's) and WAL mode, it should be possible for multiple processes to read and write to the database at the same time. There may be occassional write conflicts that we'd have to deal with by retrying, but they should be relatively rare.

However, UE's SQLite does not have the VFS support necessary to share a database between processes. The second process to try to open the database will always get an error. The best we could do would be to proceed without caching support in that second process.

**Comment by kring** [comment]
# Multiple VFSs

SQLite supports multiple VFSs, and even allows you to specify which one to use _per connection_. So, in theory, we should be able to use the platform default VFS even with the UE SQLite _for our database only_. But it'll take a little doing, because the default VFS is usually initialized in `sqlite_os_init`, which UE replaces. And it uses a bunch of platform-specific functions, all of which are static. But we could probably find a not-too-horrible way to modify sqlite3.c to make the default VFS registration callback even in the context of UE. For example, rename `sqlite3_os_init` to something else that won't conflict with anything in UE, and add a new `sqlite3_os_init` that calls the renamed function.

Of course, this still assumes we're running on a platform that has built-in SQLite VFS support. Windows, Linux, and macOS certainly do, but I don't know about iOS and Android.

**Comment by kring** [comment]
# A not-too-terrible design that uses UE's SQLite

Other than swapping in the default VFS (see above), I don't see any way to make cache database access from multiple processes work correctly using UE's SQLite. But as long as we're ok with the second Cesium for Unreal process _getting no caching at all_, I think there's a way to design a decent cache system around UE's SQLite.

The basic idea is to move all cache lookups, cache writes, and _initiation_ of HTTP requests to a dedicated cache thread. That cache owns the (one and only) connection to the cache DB. A new asset request will first be queued to the cache thread where it will check the cache database. On a miss, it will initiate a new HTTP request. When the response is returned, it will again be queued to the cache thread to be stored in the database. No synchronization is needed for database access because it all happens from a dedicated thread.

**Comment by kring** [comment]
Oh this just gets better and better:

* In Shipping builds, we're _forced_ to use UE's sqlite3 because it gets linked into the same executable as our code.
* In non-shipping builds, we _can't_ use UE's sqlite3 because it's in the `Private` subdirectory and its symbols aren't exported from the `SQLiteCore` plugin.

I wonder if it's possible to keep Cesium for Unreal in a separate .dll/.so even in shipping builds... 🤔

**Comment by ljason1993** [comment]
@kring 
another crash😭
in Development builds
https://community.cesium.com/t/maybe-bug-about-use-unreal-built-in-plugins-sqlite-open-multiple-exe-will-crash/13788/3

**Comment by kring** [comment]
I'm 90% sure this is fixed. We just need to test it thoroughly to confirm.

**Comment by kring** [comment]
I did some testing and can no longer reproduce this, so I'm closing it. Please reopen if you think you're seeing something like it again!
